src/utils/registry.ts

```diff
@@ -0,0 +1,272 @@
+    'mac11': '%s/builds/webkit/%s/webkit-mac-10.15.zip',
+    'mac11-arm64': '%s/builds/webkit/%s/webkit-mac-11.0-arm64.zip',
+    'win32': '%s/builds/webkit/%s/webkit-win64.zip',
+    'win64': '%s/builds/webkit/%s/webkit-win64.zip',
+  },
+  ffmpeg: {
+    'ubuntu18.04': '%s/builds/ffmpeg/%s/ffmpeg-linux.zip',
+    'ubuntu20.04': '%s/builds/ffmpeg/%s/ffmpeg-linux.zip',
+    'mac10.13': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac10.14': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac10.15': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac11': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac11-arm64': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'win32': '%s/builds/ffmpeg/%s/ffmpeg-win32.zip',
+    'win64': '%s/builds/ffmpeg/%s/ffmpeg-win64.zip',
+  },
+};
+
+export const allBrowserNames: BrowserName[] = Object.keys(DOWNLOAD_URLS) as BrowserName[];
+
+export const hostPlatform = ((): BrowserPlatform => {
+  const platform = os.platform();
+  if (platform === 'darwin') {
+    const [major, minor] = execSync('sw_vers -productVersion', {
+      stdio: ['ignore', 'pipe', 'ignore']
+    }).toString('utf8').trim().split('.').map(x => parseInt(x, 10));
+    let arm64 = false;
+    // BigSur is the first version that might run on Apple Silicon.
+    if (major >= 11) {
+      arm64 = execSync('sysctl -in hw.optional.arm64', {
+        stdio: ['ignore', 'pipe', 'ignore']
+      }).toString().trim() === '1';
+    }
+    // We do not want to differentiate between minor big sur releases
+    // since they don't change core APIs so far.
+    const macVersion = major === 10 ? `${major}.${minor}` : `${major}`;
+    const archSuffix = arm64 ? '-arm64' : '';
+    return `mac${macVersion}${archSuffix}` as BrowserPlatform;
+  }
+  if (platform === 'linux') {
+    const ubuntuVersion = getUbuntuVersionSync();
+    if (parseInt(ubuntuVersion, 10) <= 19)
+      return 'ubuntu18.04';
+    return 'ubuntu20.04';
+  }
+  if (platform === 'win32')
+    return os.arch() === 'x64' ? 'win64' : 'win32';
+  return platform as BrowserPlatform;
+})();
+
+export const registryDirectory = (() => {
+  const envDefined = getFromENV('PLAYWRIGHT_BROWSERS_PATH');
+  if (envDefined === '0')
+    return path.join('..', '..', '.local-browsers');
+  if (envDefined)
+    return path.resolve(process.cwd(), envDefined);
+  return path.join(cacheDirectory(), 'ms-playwright');
+})();
+
+export function isBrowserDirectory(browserDirectory: string): boolean {
+  const baseName = path.basename(browserDirectory);
+  for (const browserName of allBrowserNames) {
+    if (baseName.startsWith(browserName + '-'))
+      return true;
+  }
+  return false;
+}
+
+export class Registry {
+  private _descriptors: BrowserDescriptor[];
+
+  constructor(packagePath: string) {
+    const browsersJSON = JSON.parse(fs.readFileSync(path.join(packagePath, 'browsers.json'), 'utf8'));
+    this._descriptors = browsersJSON['browsers'];
+  }
+
+  _browser(browserName: BrowserName): BrowserDescriptor {
```

Comment:
private

Reply:
This got inlined since it shouldn't actually throw in `shouldDownload()` call